File: src/main/java/lv/ldn/java8/Crawler.java
package lv.ldn.java8;

import java.net.URL;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class Crawler {

    protected abstract String download(URL url);

    protected abstract List<URL> findLinks(String html);

    public void crawl(List<URL> links, Consumer<Stream<URL>> consumer) {
        List<CompletableFuture<List<URL>>> futures = links.stream()
                .map(url -> CompletableFuture.supplyAsync(() -> download(url)))
                .map(future -> future.thenApply(this::findLinks))
                .collect(Collectors.toList());
        CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]))
                .thenApply(v -> futures.stream()
                        .flatMap(future -> future.join().stream()))
                .thenAccept(consumer);
    }

}

File: src/main/scala/lv/ldn/scala/Crawler.scala
package lv.ldn.scala

import java.net.URL
import concurrent.ExecutionContext.Implicits.global
import concurrent.Future
import language.postfixOps

trait Crawler {
  def download(url: URL): String

  def findLinks(html: String): List[URL]

  def crawl(links: List[URL], consumer: (List[URL]) => Any) {
    val futures = links map { link =>
      Future {
        download(link)
      } map {
        findLinks
      }
    }
    Future.sequence(futures).map(_.flatMap(l => l)).foreach(list => consumer(list))
  }
}
